src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs

```diff
@@ -725,6 +731,61 @@ private async Task SendMessageAndListenAndReportTestResultsAsync(string messageT
             this.testPlatformEventSource.TranslationLayerExecutionStop();
         }
 
+        private async Task SendMessageAndListenAndReportFinalizationResultAsync(ICollection<AttachmentSet> attachments, IMultiTestRunFinalizationEventsHandler eventHandler, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var payload = new MultiTestRunFinalizationPayload
+                {
+                    Attachments = attachments
+                };
+                this.communicationManager.SendMessage(MessageType.MultiTestRunFinalizationStart, payload);
+                var isMultiTestRunFinalizationComplete = false;
+
+                using (cancellationToken.Register(() => this.communicationManager.SendMessage(MessageType.MultiTestRunFinalizationCancel)))
+                {
+                    // Cycle through the messages that the vstest.console sends.
+                    // Currently each of the operations are not separate tasks since they should not each take much time.
+                    // This is just a notification.
+                    while (!isMultiTestRunFinalizationComplete)
+                    {
+                        var message = await this.TryReceiveMessageAsync().ConfigureAwait(false);
+
+                        if (string.Equals(MessageType.MultiTestRunFinalizationComplete, message.MessageType))
+                        {
+                            if (EqtTrace.IsInfoEnabled)
+                            {
+                                EqtTrace.Info("VsTestConsoleRequestSender.SendMessageAndListenAndReportAttachments: Process complete.");
+                            }
+
+                            var multiTestRunFinalizationCompletePayload = this.dataSerializer.DeserializePayload<MultiTestRunFinalizationCompletePayload>(message);
+
+                            eventHandler.HandleMultiTestRunFinalizationComplete(multiTestRunFinalizationCompletePayload.Attachments);
+                            isMultiTestRunFinalizationComplete = true;
+                        }
+                        else if (string.Equals(MessageType.TestMessage, message.MessageType))
+                        {
+                            var testMessagePayload = this.dataSerializer.DeserializePayload<TestMessagePayload>(message);
+                            eventHandler.HandleLogMessage(testMessagePayload.MessageLevel, testMessagePayload.Message);
+                        }
```

Comment:
> } [](start = 23, length = 2)

Should we log if we get a message we cannot recognize?

Reply:
we can but this is not current pattern. Discovery and test execution are implemented in same way

Reply:
I understand, perhaps we could fix that for the new code we add in. Doesn't seem like a risky change and would help for troubleshooting. Ideally we should be throwing.

Reply:
Looking how other stuff is implemented I think throwing in this case is dangerous. I think still we can get some late messages from test run or discovery. I prefer log error in this case and send this error back through EventsHandler. WDYT?

Reply:
yup, logging was what I was recommending too.

Reply:
warning added